Pervert e2e shell tests to run under Ginkgo

zmerlynn opened the pull request:
I'm tired of people bitching about the TAP output, let's just rip the bandaid and twist this around.

cc @filbranden @satnam6502 etc.

zmerlynn committed:
Run shell tests under Ginkgo, changing all reporting to JUnit

* Add a test/e2e/shell.go that slurps up everything in hack/e2e-suite
and runs it as a bash test, borrowing all the code from hack/e2e.go.

* Rip out all the crap in hack/e2e.go that deal with multiple tests

* Move hack/e2e-suite/goe2e.sh to hack/ginkgo-e2e.sh so that it
doesn't get slurped up.

zmerlynn committed:
Enable ginkgo.v (console output is missing spec without it)

zmerlynn commented:
Oh, and not that you probably would, but please don't merge this until morning. I need to make Jenkins changes.

filbranden commented on test/e2e/shell.go:
```diff
@@ -0,0 +1,110 @@
+	return
+}
+
+func bashWrapTest(cmd string) string {
```
Noooooo!!!

Seriously though, I'm working on another PR to make the hack/e2e-suite/*.sh stand on their own so this snippet will not be required then. I'd suggest holding this one until that one is in.

filbranden commented on hack/ginkgo-e2e.sh:
```diff
@@ -89,5 +89,7 @@ fi
   --host="https://${KUBE_MASTER_IP-}" \
   --provider="${KUBERNETES_PROVIDER}" \
+  --ginkgo.v \
+  ${E2E_FOCUS+"--ginkgo.focus=${E2E_FOCUS}"} \
```
Please don't use `--ginkgo.focus`, instead use `-t` or `--test` which matches a whole test name. In fact, goe2e.sh already takes $@ arguments so you could simply pass them through here, you don't really need an exported environment variable here.

filbranden commented:
I'm not sure I agree in general... Personally I think this tends to be a way to extend the lives of those scripts and I'd rather just kill them all... But I agree that shrinking hack/e2e.go might be worth it a bit.

In any case, let's not propagate the abomination that `bashWrapTest` is. I'm working on killing it, if you can hold it a bit that would simplify this shell.go a lot.

Cheers,
Filipe

zmerlynn commented:
How are you proposing to eliminate bashWrap? You can always just eliminate
it in this code, too, it's really not that tricky to do in one more place.

Also, I'm not sure why you like -t. I very nearly deleted it outright.
Focus is such a better interface in general for spec selection?

On Wed, Feb 4, 2015, 17:31 Filipe Brandenburger <notifications@github.com>
wrote:

> I'm not sure I agree in general... Personally I think this tends to be a
> way to extend the lives of those scripts and I'd rather just kill them
> all... But I agree that shrinking hack/e2e.go might be worth it a bit.
>
> In any case, let's not propagate the abomination that bashWrapTest is.
> I'm working on killing it, if you can hold it a bit that would simplify
> this shell.go a lot.
>
> Cheers,
> Filipe
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4145#issuecomment-72978006>
> .
>

zmerlynn commented:
Also, as to the philosophical disagreement. I don't think we're going to finish porting the tests over that quickly, and right now there's an active issue in that the reporting is still muddy. Unless you have a PR that's really ready for the `bashWrap` piece, too, I'm not sure see waiting on that side, either, since it's just a tiny piece of code to modify on your PR.

filbranden commented:
I was thinking about this one a bit more. I can see the value in getting rid of TAP and moving towards JUnit.xml. However, I'm not sure what this will make to test output in case of errors. Right now it's already fairly challenging to figure out what is going on when a test fails, what will happen once *all* the tests will output to the same goe2e output?

zmerlynn commented:
I'm confused by the question. This PR elimates the TAP output completely, which is why I need to make Jenkins-side configuration changes (a minor one, to tell it to stop consuming the *.tap files). There is no `goe2e.sh` test anymore after this PR - it doesn't exist. I moved it to `hack/ginkgo-e2e.sh`. That script gets executed directly by `hack/e2e.go` as if it were just another up or down thing that `hack/e2e.go` did, like `kube-up.sh`, making `hack/e2e.go` really have no knowledge of "tests" anymore.

filbranden commented:
My point is:

Right now, if say `pd.sh` fails it's pretty easy for @satnam6502 to dig into Jenkins console and figure out where the `pd.sh` failure output is to see if it's a known problem or not.

When it's all merged into `goe2e.sh` how will that work? Will we have good logs on Jenkins? Will they all be mixed up? Will it be easy to tell the output of different failing tests apart?

satnam6502 commented:
I am at a PC meeting on Thu + Fri which is going to totally absorb every minute of my time. It's best for someone else to be assigned to this PR.

zmerlynn commented:
Sorry, I must not have explained this PR very well if you're asking this
question. Every shell test ends up as a Gingko spec. I borrowed the code
for capturing stdout/stderr. As readable in the code, when the shell test
falls, it gets recorded as a Gingko Fail with the stdout/stderr, the same
way we used to do TAP output (well, modulo the crappy "YAMLish" escaping).

I'll share some failure output when I get back to a real computer.
On Feb 4, 2015 7:58 PM, "Filipe Brandenburger" <notifications@github.com>
wrote:

> My point is:
>
> Right now, if say pd.sh fails it's pretty easy for @satnam6502
> <https://github.com/satnam6502> to dig into Jenkins console and figure
> out where the pd.sh failure output is to see if it's a known problem or
> not.
>
> When it's all merged into goe2e.sh how will that work? Will we have good
> logs on Jenkins? Will they all be mixed up? Will it be easy to tell the
> output of different failing tests apart?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/4145#issuecomment-72990364>
> .
>

zmerlynn commented:
I updated the commit minorly to include "By()" output of the actual .sh getting run, and fixed a somewhat glaring bug. I'm now including output. This is using a simple e2e test called silly.sh that looks like:

```
echo Normal output
echo  more normal output
if (( $RANDOM % 2 == 0 )); then
  2>& echo SOME ERROR
  exit 1
fi
```

Here's a sample failure:

```
$ E2E_FOCUS=silly.sh go run hack/e2e.go -test -v -check_version_skew=false
2015/02/04 23:31:03 Running in background: watchEvents
2015/02/04 23:31:03 Running: get status
Project: artful-fortress-786
Zone: us-central1-f
Running: /usr/local/google/home/zml/kubernetes/cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl version
Client Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-87-g750973597af82a-dirty", GitCommit:"750973597af82ae9e83defcc86c7a21d3e6d88c8", GitTreeState:"dirty"}
Server Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-86-ge6e67e26625efc-dirty", GitCommit:"e6e67e26625efc2a3c32ed2c9371da94bc8e3c9d", GitTreeState:"dirty"}
2015/02/04 23:31:05 Running: validate cluster size
Project: artful-fortress-786
Zone: us-central1-f
Running: cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl get minions --no-headers
2
2015/02/04 23:31:08 Running: Ginkgo tests
Project: artful-fortress-786
Zone: us-central1-f
Project: artful-fortress-786
Zone: us-central1-f
Running Suite: Kubernetes e2e Suite
===================================
Random Seed: 1423121470
Will run 1 of 20 specs

SSSSSS
------------------------------
Shell
  tests that silly.sh passes
  /usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:48
STEP: Running /usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh

• Failure [0.506 seconds]
Shell
/usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:50
  tests that silly.sh passes [It]
  /usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:48

  Error running /usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh:
  stdout:
  Normal output
  more normal output 
 
  stderr:
  Project: artful-fortress-786
  Zone: us-central1-f 
  /usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh: line 6: echo: ambiguous redirect



  /usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:85
------------------------------
SSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] Shell [It] tests that silly.sh passes
/usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:85

Ran 1 of 20 Specs in 0.507 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 19 Skipped F0204 23:31:11.425001   24527 driver.go:88] At least one test failed
2015/02/04 23:31:11 Error running Ginkgo tests: exit status 255
watchEvents stderr: | 
  Project: artful-fortress-786
  Zone: us-central1-f 
  Running: /usr/local/google/home/zml/kubernetes/cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl --watch-only get events
```

It's kind of chatty because the failure is embedded in the Gingko, but if you're looking at the console logs, you kind of get what you get. The JUnit will, of course, have something closer to just the multi-line failure message in the middle.

A success is shorter:
```
$ E2E_FOCUS=silly.sh go run hack/e2e.go -test -v -check_version_skew=false
2015/02/04 23:39:11 Running in background: watchEvents
2015/02/04 23:39:11 Running: get status
Project: artful-fortress-786
Zone: us-central1-f
Running: /usr/local/google/home/zml/kubernetes/cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl version
Client Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-87-g750973597af82a-dirty", GitCommit:"750973597af82ae9e83defcc86c7a21d3e6d88c8", GitTreeState:"dirty"}
Server Version: version.Info{Major:"0", Minor:"10+", GitVersion:"v0.10.0-86-ge6e67e26625efc-dirty", GitCommit:"e6e67e26625efc2a3c32ed2c9371da94bc8e3c9d", GitTreeState:"dirty"}
2015/02/04 23:39:14 Running: validate cluster size
Project: artful-fortress-786
Zone: us-central1-f
Running: cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl get minions --no-headers
2
2015/02/04 23:39:17 Running: Ginkgo tests
Project: artful-fortress-786
Zone: us-central1-f
Project: artful-fortress-786
Zone: us-central1-f
Running Suite: Kubernetes e2e Suite
===================================
Random Seed: 1423121959
Will run 1 of 20 specs

SSSSSSSSSSSSSSSSS
------------------------------
Shell 
  tests that silly.sh passes
  /usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:48
STEP: Running /usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh
•SS
Ran 1 of 20 Specs in 0.557 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 19 Skipped I0204 23:39:20.077577   27185 driver.go:90] All tests pass
watchEvents stderr: |
  Project: artful-fortress-786
  Zone: us-central1-f
  Running: /usr/local/google/home/zml/kubernetes/cluster/../cluster/gce/../../_output/local/bin/linux/amd64/kubectl --watch-only get events
  
exit status 1
```

Note that a lot of that chatter was common `hack/e2e.go` chatter anyways. (Also note why I was using E2E_FOCUS here, but obviously I didn't need to be going through hack. I was going through hack anyways, because I wanted to make sure things were roughly sane. And obviously I've run full runs, too, but I do most of my testing on a rinky-dink test like that.)

Does that make it clearer?

zmerlynn commented:
Here's the silly.sh xml snippet from the JUnit (totally non-helpful for a human, as is most JUnit XML):

```
      <testcase name="Shell tests that silly.sh passes" classname="Kubernetes e2e Suite" time="0.580071756">
          <failure type="Failure">/usr/local/google/home/zml/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:48&#xA;Error running /usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh:&#xA;stdout:&#xA;Normal output&#xA;more normal output&#xA;&#xA;stderr:&#xA;Project: artful-fortress-786&#xA;Zone: us-central1-f&#xA;/usr/local/google/home/zml/kubernetes/hack/e2e-suite/silly.sh: line 6: echo: ambiguous redirect&#xA;&#xA;</failure>
      </testcase>
```

zmerlynn commented:
Here's an example full run, in JUnit:

```
$ less junit.xml 
<?xml version="1.0" encoding="UTF-8"?>
  <testsuite tests="19" failures="0" time="1218.926633819">
      <testcase name="TestBasic should pass" classname="Kubernetes e2e Suite" time="41.331934756"></testcase>
      <testcase name="TestPrivate should pass" classname="Kubernetes e2e Suite" time="47.357495641"></testcase>
      <testcase name="TestKubeletSendsEvent should pass" classname="Kubernetes e2e Suite" time="10.762725963"></testcase>
      <testcase name="Shell tests that certs.sh passes" classname="Kubernetes e2e Suite" time="9.444929877"></testcase>
      <testcase name="Shell tests that guestbook.sh passes" classname="Kubernetes e2e Suite" time="30.694180711"></testcase>
      <testcase name="Shell tests that liveness.sh passes" classname="Kubernetes e2e Suite" time="165.433108971"></testcase>
      <testcase name="Shell tests that monitoring.sh passes" classname="Kubernetes e2e Suite" time="290.925709228"></testcase>
      <testcase name="Shell tests that pd.sh passes" classname="Kubernetes e2e Suite" time="136.320303995"></testcase>
      <testcase name="Shell tests that services.sh passes" classname="Kubernetes e2e Suite" time="159.980516205"></testcase>
      <testcase name="Shell tests that update.sh passes" classname="Kubernetes e2e Suite" time="132.139362395"></testcase>
      <testcase name="TestKubernetesROService should pass" classname="Kubernetes e2e Suite" time="0.48774776000000003"></testcase>
      <testcase name="TestLivenessHttp should pass" classname="Kubernetes e2e Suite" time="26.220678453"></testcase>
      <testcase name="TestClusterDNS should pass" classname="Kubernetes e2e Suite" time="6.067550859"></testcase>
      <testcase name="Pods should be submitted and removed" classname="Kubernetes e2e Suite" time="0.807316268"></testcase>
      <testcase name="Pods should be updated" classname="Kubernetes e2e Suite" time="82.435981103"></testcase>
      <testcase name="Pods should contain environment variables for services" classname="Kubernetes e2e Suite" time="16.686152809"></testcase>
      <testcase name="TestImportantURLs should pass" classname="Kubernetes e2e Suite" time="0.36880817800000004"></testcase>
      <testcase name="TestNetwork should pass" classname="Kubernetes e2e Suite" time="22.874984253"></testcase>
      <testcase name="TestLivenessExec should pass" classname="Kubernetes e2e Suite" time="38.584585706"></testcase>
  </testsuite>
```

zmerlynn commented:
Rebase incoming after #4156.

zmerlynn commented:
@filbranden: PTAL. I rebased, removed the E2E_FOCUS commit, changed the Jenkins script over to just use the a simple `export KUBE_CONFIG_FILE="config-test.sh"` method versus touching hack/e2e (this works just fine, and used to before your commit too .. I was considering making this change earlier and your commit prompted me that it was about time).

I did test that `hack/e2e.go -test` still works for the dev workflow, but I wanted to take it out of the Jenkins flow for now to give us a little cover. (But people can also just call `hack/ginkgo-e2e.sh` directly after this goes in.) Now we should just be able to rip up hack/e2e.go with a little more impunity.

I'd like to get this in early today if possible so that we get a good JUnit history going, because there's some actually aggregation packages available for the JUnit side. (Plus it means I can catch fallout early in the day.)

filbranden commented:
Looks great! Happy to see we're using exec.Command to run the bash scripts now!

filbranden commented:
Ok I do see one blocker here... You need to add hack/ginkgo-e2e.sh to KUBE_TEST_PORTABLE in hack/lib/golang.sh, I'm almost sure this will break Jenkins if you don't... Can you update the PR to add a third commit for it?

Thanks!
Filipe

zmerlynn commented:
Oh, hey, hoisted by my own petard. I forgot to run `e2e-from-release.sh` after setting all that up. Will add and double check that works.

zmerlynn commented:
(Good catch)

zmerlynn committed:
Add hack/ginkgo-e2e.sh to kubernetes-test.tgz

filbranden commented:
I might be away from keyboard on the next hour... so feel free to self-merge when this is ready.

BTW, this will also break `cluster/test-network.sh` but I'd rather you *not* fix it since there's really no advantage in running that instead of `hack/ginkgo-e2e.sh -t TestNetwork`, maybe you'd want to remove that script on this PR? Either remove it, fix it or leave it broken, but if not removed now I'll remove it next...

zmerlynn committed:
Remove cluster/test-network.sh

To fix this, it would be an alias for:

  hack/ginkgo-e2e.sh -t TestNetwork

... which may not work on binary distributions, because it's relying
on test assets, so it probably shouldn't live in cluster/ ...

zmerlynn commented:
Cool, I went ahead and removed it with justifcation, and reopened #2178.

zmerlynn commented:
I'll merge on Travis.
